airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -410,22 +409,39 @@ def read_records(
             slice_state = {}
 
         most_recent_record_from_slice = None
-        for record in self._read_pages(self.parse_records, stream_slice, slice_state):
-            if self.cursor:
-                if not most_recent_record_from_slice:
-                    most_recent_record_from_slice = record
-                else:
-                    most_recent_record_from_slice = (
-                        most_recent_record_from_slice
-                        if self.cursor.is_greater_than_or_equal(most_recent_record_from_slice, record)
-                        else record
-                    )
-            yield record
+        for stream_data in self._read_pages(self.parse_records, stream_slice, slice_state):
+            most_recent_record_from_slice = self._get_most_recent_record(most_recent_record_from_slice, stream_data, stream_slice)
+            yield stream_data
 
         if self.cursor:
             self.cursor.close_slice(stream_slice, most_recent_record_from_slice)
         return
 
+    def _get_most_recent_record(
+        self, current_most_recent: Optional[Record], stream_data: StreamData, stream_slice: StreamSlice
+    ) -> Optional[Record]:
+        if self.cursor and (record := self._extract_record(stream_data, stream_slice)):
+            if not current_most_recent:
+                return record
+            else:
+                return current_most_recent if self.cursor.is_greater_than_or_equal(current_most_recent, record) else record
+        else:
+            return None
+
+    @staticmethod
+    def _extract_record(stream_data: StreamData, stream_slice: StreamSlice) -> Optional[Record]:
+        """
+        As we allow the output of _read_pages to be StreamData, it can be multiple things. Therefore, we need to filter out and normalize
+        to data to streamline the rest of the process.
+        """
+        if isinstance(stream_data, Record):
+            # Record is not part of `StreamData` but is the most common implementation of `Mapping[str, Any]` which is part of `StreamData`
```

Comment:
is there a reason `Record` shouldn't be part of `StreamData`?

Reply:
This is a good question. I would tend to try to remove things from `StreamData` instead of adding some. I honestly hate the fact that we need to check the type of the return object when calling a method so I would try to avoid adding to this problem. Personally, I would love to have `HttpStream._read_pages` return records but I haven't evaluate the work required to do this. As we allow for `CustomRetriever`, it might be harder but at the same time, I don't see any in airbyte-integrations